Issue #18 upload config file

andrewmadden opened the pull request:
* Validates uploaded file if it finds one. 
* Only shows file uploader and quit password if seb required setting is set to upload config.

> Quit password is stored as a hash in config file, so can't be extracted into the setting.

andrewmadden committed:
Issue #18 Enable upload config form options

andrewmadden committed:
Issue #18 Validate uploaded config file

andrewmadden committed:
Issue #18 Fix settings provider unit test

andrewmadden committed:
Issue #18 Add error string for file parse failure

andrewmadden committed:
Issue #18 Add unit tests to cover file validation

nhoobin committed:
Issue #18 Converting filepicker to filemanager in settings provider

nhoobin committed:
Issue #18 Update the form definition for the filemanager type exception

nhoobin committed:
Issue #18 Update the form definition validation for the filemanger

nhoobin committed:
Issue #18 Language file update for the filemanger

nhoobin committed:
Issue #18 Save the form filemanager upload into the module context filearea

nhoobin committed:
Issue #18 Unit text fixes for the filemanager changes

nhoobin committed:
Issue #18 Ignore line for outdated PHPCS

nhoobin committed:
Issue #18 Adding backup / restore for the filemanager files

nhoobin committed:
Issue #18 Create reference to module context file area id for the config

nhoobin committed:
Issue #18 Update calls to filestorage commands to use the new const

const SEB_CONFIG_FILE_ITEMID = 111;

dmitriim commented on classes/settings_provider.php:
```diff
@@ -58,6 +58,11 @@ class settings_provider {
+     /**
+      * The filearea itemid which is used for saving the module template.
+     */
+    const SEB_CONFIG_FILE_ITEMID = 111;
```
Why is it 111?

nhoobin replied:
To remove the magic numbers that were littered throughout the code base. 111 could just be 1.
Specifically not 0 as I didn't want the file itemid to be equated to 0/false anywhere.

dmitriim replied:
Maybe pick generic name for that not related to seb config file as it could change in the future. Why not to use $cm->id? Not saying that it should be, just asking...

dmitriim commented on classes/settings_provider.php:
```diff
@@ -58,6 +58,11 @@ class settings_provider {
     const USE_SEB_CLIENT_CONFIG = 4;
 
+     /**
+      * The filearea itemid which is used for saving the module template.
```
Is this doc correct?  are we using it for template?

nhoobin replied:
Nice one, I'll adjust that.

dmitriim commented on classes/settings_provider.php:
```diff
@@ -70,7 +75,7 @@ public static function get_quiz_element_types() : array {
             'seb_requiresafeexambrowser' => ['select', self::get_requiresafeexambrowser_options()],
             'seb_sebconfigtemplate' => ['select', self::get_template_options()],
-            'seb_sebconfigfile' => 'filepicker',
+            'filemanager_sebconfigfile' => ['filemanager', self::get_filemanager_options()],
```
Is name important here for moodle API? If not I'd suggest to rename for consistency, as all other are like seb_***

nhoobin replied:
The name was changed here as to not conflict with the existing quiz_settings persistent property 'sebconfigfile'.

Once the hook to add modform elements has finished, a $mform->set_data(...) function is called and overwrites the form element 'seb_sebconfigfile'. Which then prevented displaying the saved config file from the module context.

nhoobin replied:
Also the mod_form hook for quiz access plugins did not allow for other moodle form hooks to be called, eg, data_preprocessing, definition_after_data etc.
```
mod_form.php:346
// Any other rule plugins.
quiz_access_manager::add_settings_form_fields($this, $mform);
```
With this, setting the data with a set_data() call was failing because the form was not even constructed completely at that time. 

Given that seb_configfile was being saved, and was part of the persistent class, when the saving the modform data, an eventual setdata call was made and it just restored the users draft area, or no items at all.

To get around this I made this exception. Though I'd love to know how we can keep it generic and not require this at all.

dmitriim commented on rule.php:
```diff
@@ -114,6 +116,19 @@ public static function add_settings_form_fields(mod_quiz_mod_form $quizform, Moo
+                    $quizform->get_context()->id,
+                    'quizaccess_seb',
+                    $name,
+                    settings_provider::SEB_CONFIG_FILE_ITEMID
```
This code supposed to be flexible and could be used for any filemanager in the form. I'd suggest to get rid of settings_provider::SEB_CONFIG_FILE_ITEMID as it adds some confusion. Is it safe to hardcode to 0? Seems like it could be anything and different parts of moodle use different value :thinking:

dmitriim commented on tests/phpunit/quiz_settings_test.php:
```diff
@@ -167,6 +168,39 @@ public function test_browser_exam_keys_validation_errors($bek, $expectederrorstr
+    public function test_validate_sebconfigfile_success() {
+        $user = $this->getDataGenerator()->create_user();
+        $this->setUser($user);
+        $xml = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
```
is it possible to use file_get_contents(__DIR__ . '/sample_data/simpleunencrypted.seb'); here?

nhoobin replied:
Correct, the unit tests could have fixtures which are loaded like that. Yet those are not real seb files themselves.

dmitriim commented on tests/phpunit/quiz_settings_test.php:
```diff
@@ -167,6 +168,39 @@ public function test_browser_exam_keys_validation_errors($bek, $expectederrorstr
+        $this->assertEmpty($errors);
+    }
+
+    public function test_validate_sebconfigfile_failure() {
```
Nice that you added negative scenario test as well :+1:

nhoobin replied:
Thats all @andrewmadden  :tada:

dmitriim commented on rule.php:
```diff
@@ -155,6 +170,10 @@ public static function validate_settings_form_fields(array $errors,
         $quizsettings->set('quizid', $data['instance']);
+        // Explicit setting the sebconfigfile value such that the validation will check for a valid plist.
+        if (isset($data['filemanager_sebconfigfile'])) {
+            $quizsettings->set('sebconfigfile', $data['filemanager_sebconfigfile']);
```
in DB table quizaccess_seb_quizsettings "sebconfigfile' and persistent class is text. Do you think we need to change it?

nhoobin replied:
This was my hack around to keep that value in the persistent class, and it will store the itemid to acknowledge that the file is been uploaded. In conjunction with the workarounds I mentioned above.

dmitriim commented on classes/quiz_settings.php:
```diff
@@ -402,4 +438,49 @@ private function split_keys(string $keys) : array {
+        global $USER;
+        $context = context_user::instance($USER->id);
+        $fs = get_file_storage();
+        $files = $fs->get_area_files($context->id, 'user', 'draft', $itemid);
```
In bunch of places in moodle  I saw something like 
```
if (!$files = $fs->get_area_files($context->id, 'user', 'draft', $draftid, 'id DESC', false)) {
                return false;
}
 $file = reset($files);
```
They don't include dirs and change sort order to get the latest file.  Also no check for file size. Just picking the latest.
Do you think we need to change it here as well?

dmitriim commented on classes/quiz_settings.php:
```diff
@@ -402,4 +438,49 @@ private function split_keys(string $keys) : array {
+    private function save_filemanager_draftarea() : bool {
+        global $CFG;
+
+        // During plugin installation file_save_draft_area_files causes an exception that user id=0 is not found.
```
Do you know why this method gets called During plugin installation  ?

nhoobin replied:
During the install.php there is a function quizaccess_seb_create_missing_settings();

The lib.php:75 function creates settings for all quizzes where there are none.

dmitriim replied:
but should it return empty $draftitemid on istall?

nhoobin replied:
No files were present during the install so it is not modifying the default value.

Out of this scope I asked why the settings were generated during install at all. Can we modify the access rule constructor to not throw an exception when settings are not found?

dmitriim replied:
Sure. If you think it will make the code better, then go ahead!

dmitriim replied:
New issue created https://github.com/catalyst/moodle-quizaccess_seb/issues/88

nhoobin committed:
Issue #18 Remove unrequired check

nhoobin committed:
Issue #18 Change generic function to be specific for saving draft files

nhoobin committed:
Issue #18 Remove constant and use $cm->id for files itemid

nhoobin committed:
Issue #18 Cleaning up the get_area_files call to be more readable

nhoobin committed:
Issue #18 Remove any uploaded config seb when the mod form is saved

When settings_provider::USE_SEB_UPLOAD_CONFIG is not set, this will delete
any saved configuration file that exist in the module context.

nhoobin committed:
Issue #18 Unit test fixes

nhoobin committed:
Issue #18 Updating validation for missing / invalid SEB file content

dmitriim commented on classes/quiz_settings.php:
```diff
@@ -253,6 +296,12 @@ private function before_save() {
+        if ($this->get('requiresafeexambrowser') == settings_provider::USE_SEB_UPLOAD_CONFIG) {
+            $this->save_filemanager_sebconfigfile_draftarea();
+        } else {
+            $this->cleanup_draftareas_and_saved_config();
```
Would be good to cover this scenario with unit tests.

dmitriim commented on classes/quiz_settings.php:
```diff
@@ -253,6 +296,12 @@ private function before_save() {
         $this->compute_config_key();
+
+        if ($this->get('requiresafeexambrowser') == settings_provider::USE_SEB_UPLOAD_CONFIG) {
+            $this->save_filemanager_sebconfigfile_draftarea();
```
Also would be good to cove this with the tests. It could be tricky, but in test we can hack it somehow or refactor code to make it more testable. E.g. move getting draftitemid outside of this class.

dmitriim commented on classes/quiz_settings.php:
```diff
@@ -402,4 +451,67 @@ private function split_keys(string $keys) : array {
+     * @return bool Always true or exception if error occurred
+     * @throws \coding_exception
+     */
+    private function cleanup_draftareas_and_saved_config() : bool {
```
how about to rename this method? Maybe delete_uploaded_config_file()?

dmitriim commented on classes/quiz_settings.php:
```diff
@@ -402,4 +451,67 @@ private function split_keys(string $keys) : array {
+     * @throws \coding_exception
+     */
+    private function save_filemanager_sebconfigfile_draftarea() : bool {
+        $draftitemid = file_get_submitted_draft_itemid('filemanager_sebconfigfile');
```
let's refactor this and move file uploading to the form processing part of the code. This will decouple persistent from the form submitting. I recon we can just pass file/fileid to persistent

nhoobin committed:
Issue #18 Refactoring out file saving logic from persistent class to helper

dmitriim commented on classes/quiz_settings.php:
```diff
@@ -208,6 +215,42 @@ protected function validate_allowedbrowserexamkeys($keys) {
+            return true;
+        }
+        // If there is a config file uploaded, make sure it is a PList XML file.
+        $file = settings_provider::get_current_user_draft_file($itemid);
```
is it possible to move this out of the class as well? This is very coupled with submitting the form which I'd like to avoid.

nhoobin committed:
Issue #18 Refactor file validation out of persistent class to helper

nhoobin committed:
Issue #18 Removing test for persistent class to validate file

The file should be added via a form element and referenced in the file
storage with a $cmid as one of its keys.

nhoobin committed:
Issue #18 Fixing autoloading in settings_provider for using property list

nhoobin committed:
Issue #18 Restoring tests for validating form uploaded config file.

nhoobin committed:
Issue #18 phpcs bugfix

nhoobin committed:
Issue #18 Removing unnecessary validation after the refactor

nhoobin committed:
Issue #18 String fix

nhoobin committed:
Issue #18 update params for save_filemanager_sebconfigfile_draftarea

Now allowing to pass through the file_get_submitted_draft_itemid for unit
testing

nhoobin committed:
Issue #18 Updating unit tests for settings_provider
